### PR TITLE
feat: Add cascading db for dry runs

### LINKED ISSUE #114: Bug in validate call
Validate only works for blocks that are already in the db but it's purpose is dry running so we need a way to put the block "in the db" but without actually putting it in there. Normally we would use a transaction and just not commit it but we can't do that due to async. We could create an in memory db and stick it in that and then fallback to the disk db if no state is found.

### PR DESCRIPTION
**Pull request by freesig**
Dry running validate requires the block be in the database but we don't want to put the dry run block in the actual database. 
This allows adding the block to an in-memory database that cascades to the on-disk db if no state is found.
fixes #114

### CODE REVIEW

File: crates/node/src/validate.rs [hunk 1]
```diff
@@ -27,7 +27,53 @@ struct State {
...
+pub struct DryRun {
+    memory: Memory,
+    conn_pool: ConnectionPool,
+}
+
+#[derive(Clone)]
+/// In-memory database that contains a dry run block.
+struct Memory {
+    memory: db::ConnectionPool,
+}
```
Comment (supiket) [inline crates/node/src/validate.rs:53]:
```suggestion
struct Memory(db::ConnectionPool)
```

File: crates/node/src/validate.rs [hunk 2]
```diff
@@ -110,6 +154,33 @@ pub async fn validate(
...
+    let dry_run = DryRun::new(conn_pool.clone(), block).await?;
+    let db_type = Db::DryRun(dry_run);
+    validate_inner(db_type, contract_registry, block).await
+}
+
+/// Validates a block.
+///
+/// Returns a `ValidationResult` if no `ValidationError` occurred that prevented the block from being validated.
+#[cfg_attr(feature = "tracing", tracing::instrument(skip_all))]
+pub(crate) async fn validate_incoming_block(
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:167]: Maybe `validate_in_db` or similar might be a better name for this? Just wondering if `validate_incoming_block` might still be a bit ambiguous. Or maybe we can mention in the docs that, despite passing the `Block` as an argument, we actually query its state from the DB?

File: crates/node/src/validate.rs [hunk 3]
```diff
@@ -208,6 +279,96 @@ pub async fn validate(
...
+            Conn::Handle(handle) => {
+                let tx = handle.transaction()?;
+                Ok(Transaction::Handle(tx))
+            }
+        }
+    }
+}
+
+/// Cascade from in-memory to on-disk database across transactions.
+fn cascade(
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:351]: At first from your PR comment I thought `cascade` was referring to using some sql cascade functionality or something, but I see that you're using it for the query fallback behaviour. I'm not sure about using cascade, but I can see where you're coming from with it.

File: crates/node/src/validate.rs [hunk 4]
```diff
@@ -27,7 +27,51 @@ struct State {
...
+enum Db {
+    DryRun(DryRun),
+    ConnectionPool(ConnectionPool),
+}
+
+#[derive(Clone)]
+/// A dry run database.
+///
+/// Cascades from in-memory to on-disk database.
+pub struct DryRun {
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:44]: Do we need to expose this type publicly? Or is it enough to have some pub `validate_dry_run` fn or similar?

File: crates/node/src/validate.rs [hunk 5]
```diff
@@ -110,6 +154,33 @@ pub async fn validate(
     conn_pool: &ConnectionPool,
     contract_registry: &ContentAddress,
     block: &Block,
+) -> Result<ValidateOutcome, ValidationError> {
+    let dry_run = DryRun::new(conn_pool.clone(), block).await?;
+    let db_type = Db::DryRun(dry_run);
+    validate_inner(db_type, contract_registry, block).await
+}
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:161]: It looks like this makes the `validate` behaviour the dry-run behaviour - should we name this `validate_dry_run` for clarity, or maybe update the docs?

**Review by mitchmindtree — COMMENTED** [review]
Something feels a little off about dynamically dispatching over the `Db`, `Conn` and `Transaction` like this, but it's hard to put my finger on it :thinking:

I guess it feels like a bit of a hack around `validate`, as a result of `validate` always assuming the next block we're trying to validate is already in the DB. This assumption makes sense in the context of the node's normal validation, but now we're finding it's making this dry-run behaviour tricky which you're aiming to solve here.

I think what's throwing me off is that this `DryRun` behaviour is basically what the builder is already doing by default, i.e. validating a block of solutions in-memory before committing or doing something else with them. That said I totally see why we want this behaviour in the Node, as it will be useful for users to be able to dry run their solutions in a new block locally before submitting them.

## Rip out builder validation?

Another option might be to rip out the validation from the builder (i.e. [`essential_builder::check_solutions`](https://github.com/essential-contributions/essential-builder/blob/1eb063bb263dc67fb37e22e1a7c79b064de859f1/crates/builder/src/lib.rs#L305) and almost everything it depends on) and move it into an `essential-validate` crate or similar in this repo? This would be possible as of #115 as `essential-validate` could depend on `essential-node-db` and still get access to the `ConnectionPool`.

The builder's validation already has parallel solution checking and predicate querying, and the [`Mutations` type](https://github.com/essential-contributions/essential-builder/blob/1eb063bb263dc67fb37e22e1a7c79b064de859f1/crates/builder/src/state/mutations.rs#L10) which is probably more efficient than spinning up a new in-memory connection pool?

Maybe a bit of a stretch, but we could potentially also make this the default way that the node performs validation in general? The node already reads out a whole block into memory for validation anyway... we could just construct the in-memory `Mutations` from that and then both node and builder would use an identical solutions validation process. I guess the cost might be slightly suboptimal node validation efficiency.

---

Anyways, I'd be happy to see this land in the meantime while we decide on sharing more parts of validation - just left a few small comments, but nothing really blocking.

**Comment by freesig** [comment]
Is the worry here anything other then performance? This is not something we currently provide and API for. If we do then it will probably be unavailable on the deployed nodes anyway so I'm not sure performance really matters or that this really is that slower then the other more complex approach.

The cascade approach allows use to use any sql queries we write in the future for free. This is equivalent to an transaction. When we write a block to the on disk db it first goes into an in memory tx. The only difference is this lives a bit longer so the memory is used a bit longer. The only other overhead is creating the tables but I really don't think that's very slow and it's constant overhead that's not linked to the block size.

The builder validation would not work for the validation stream so if we create another crate for validation it's only going to be shared between the dry run and the builder. Then we'd have to maintain two separate validation functions within the node which really need to have the same behavior.

I think unless we have evidence that this approach is drastically slow then it's not worth the extra complexity vs this very simple implementation.
The other path this leaves open is the ability to switch to a sql db on disk that gets cleaned up after validation. For very large blocks on resource constrained devices this might be pretty useful.

File: crates/node/src/validate.rs (hunk 2 above)
Comment (freesig) [reply crates/node/src/validate.rs]: Changed this to `validate`.

**Comment by mitchmindtree** [comment]
> The builder validation would not work for the validation stream

Partly what I'm suggesting is that it *could* work for the validation stream too.

We already read the whole block out just to iterate over the solutions for the node during validation - it'd be trivial to convert the current block being validated into a [`Mutations`](https://github.com/essential-contributions/essential-builder/blob/1eb063bb263dc67fb37e22e1a7c79b064de859f1/crates/builder/src/state/mutations.rs#L10) instance and then re-use all the same validation approach for solutions between the node, dry-run and builder.

> I think unless we have evidence that this approach is drastically slow then it's not worth the extra complexity vs this very simple implementation.

My concern is mostly about having two separate implementations of almost the same thing in both the builder and node. My suggestion was mostly aimed at _reducing_ complexity by avoiding adding the extra 200 lines here and doing some refactor to share validation logic with the builder.

> The other path this leaves open is the ability to switch to a sql db on disk that gets cleaned up after validation. For very large blocks on resource constrained devices this might be pretty useful.

If this is something we want to support, it does rule out using the builder's approach.

> The cascade approach allows use to use any sql queries we write in the future for free. This is equivalent to an transaction

I get the idea behind having a generalised transaction-like type around the connection pool with another in-memory connection pool and how it offers freedom to add more SQL queries more easily, I'm just not sure this is something we want to plan or optimise for when the only thing we currently use the DB for during validation of a block's solutions is querying state, and afaik we don't have any plans for adding more?

**Comment by freesig** [comment]
> Partly what I'm suggesting is that it could work for the validation stream too.

> We already read the whole block out just to iterate over the solutions for the node during validation - it'd be trivial to convert the current block being validated into a [Mutations](https://github.com/essential-contributions/essential-builder/blob/1eb063bb263dc67fb37e22e1a7c79b064de859f1/crates/builder/src/state/mutations.rs#L10) instance and then re-use all the same validation approach for solutions between the node, dry-run and builder.

This would involve creating the `Mutations` which is not needed for the validation stream. Also it would double the amount of memory for mutations because they would be pulled into memory for both the `solutions` and `mutations`. Also for really large blocks or resource constrained devices there's actually no reason to pull the entire block into memory as we could just pull batches of solutions or even parts of a `solution` if we have one big solution. If we take the `Mutations` approach then we would could build it up incrementally in memory as we validate `solutions` but you'd still need to have the whole thing in memory by the last `solution`. If we have one big `solution` then we would have to have all `Mutations` in memory to begin with.

 > My concern is mostly about having two separate implementations of almost the same thing in both the builder and node. My suggestion was mostly aimed at reducing complexity by avoiding adding the extra 200 lines here and doing some refactor to share validation logic with the builder.
 
 I don't really think this is a big problem. It's only 200 lines and could probably be condensed if we used traits.  I can look into reducing the size of this implementation. I think it's pretty reasonable to have two different implementations as the builder has pretty different priorities to dry runs. I think it's going to be pretty rare that devs will be dry running a large block. They are probably just validating a couple of solutions. If they really need the speed (say for a solver) they could use the builder directly or even write a custom function that suites them better.

**Comment by freesig** [comment]
Just added https://github.com/essential-contributions/essential-node/pull/121 as an alternative. This uses traits instead of enums. Not sure which is better

**Review by mitchmindtree — APPROVED** [review]
> Just added https://github.com/essential-contributions/essential-node/pull/121 as an alternative.

Just had a quick look and I think I could go either way - maybe the use of `enum`s here in this PR is a bit simpler in the end? Will leave the call up to you :ok_hand:

I think now that `DryRun` is private and we don't really expose any of these details publicly I feel good about seeing this land :+1: We can always experiment with consolidating some of the validation stuff in a follow-up without affecting the pub interface.

**Comment by freesig** [comment]
Issue fyi https://github.com/essential-contributions/essential-node/issues/123

**Pull request merged**